File: apps/render-ios-playground/render-ios-playground/Domain/Services/ComponentRenderingService.swift
import UIKit

/// Domain service responsible for rendering components to UIViews
class ComponentRenderingService {
    private var renderers: [String: Renderer] = [:]
    
    init(renderers: [Renderer]) {
        for renderer in renderers {
            self.renderers[renderer.type] = renderer
        }
    }
    
    
    func render(_ component: Component) -> UIView? {
        guard let renderer = renderers[component.type] else {
            return nil
        }
        
        guard let view = renderer.render(component: component) else {
            return nil
        }
        
        for child in component.getChildren() {
            if let subview = self.render(child) {
                view.addSubview(subview)
            }
        }
        
        return view
    }
}

File: apps/render-android-playground/app/src/main/java/com/max_mrtnv/render_android_playground/domain/services/ComponentRenderingService.kt
package com.max_mrtnv.render_android_playground.domain.services

import androidx.compose.runtime.Composable
import com.max_mrtnv.render_android_playground.domain.entities.Component

/**
 * Domain service responsible for rendering components to Composables
 */
class ComponentRenderingService(
    private val renderers: Map<String, Renderer>
) {
    @Composable
    fun render(component: Component): Boolean {
        val renderer = renderers[component.type] ?: return false
        return renderer.render(component)
    }
}

/**
 * Interface for component renderers
 */
interface Renderer {
    val type: String
    
    @Composable
    fun render(component: Component): Boolean
}
